shell.nix

```diff
@@ -0,0 +1,58 @@
+
+# This shell.nix provides a simple development environment for the btnl_client python package.
+#
+# First, run `nix-shell` to get into the development environment:
+#
+# $ nix-shell
+#
+# This puts you in a shell with a `python` binary on the `PATH`.  The `python` binary is
+# setup to be able to find all required python dependencies (like the
+# `requests` and `websockets` packages).
+#
+# From within this Nix shell, you can run the `bntl_client` module like the
+# following to play around with it:
+#
+# $ python -m btnl_client
+#
+# Here is an example of running one of the available subcommands.  This gets
+# the specifications for a Product:
+#
+# $ python -m btnl_client get-product-spec 3
+# ProductSpreadSpec(type='spread', product_id=3, product_name='Bitcoin US Dollar Calendar Spread', ...
+
+{...}:
+
+let
+  nixpkgs-src = builtins.fetchTarball {
+    # nixos-unstable as of 2023-08-11
+    url = "https://github.com/NixOS/nixpkgs/archive/ce5e4a6ef2e59d89a971bc434ca8ca222b9c7f5e.tar.gz";
+    # obtained with 'nix-prefetch-url --unpack <url>'
+    sha256 = "16ls4izhj8jwm8s4ydjma6ghr76xxrclcw86iy4id3bm9i60n8l1";
+  };
+
+  pkgs = import nixpkgs-src {};
+
+  python-with-my-packages =
+    pkgs.python39.withPackages
+      (python-pkgs:
+        [ # python build/test/dev dependencies:
+          python-pkgs.pytest
+          python-pkgs.build
+          python-pkgs.python-lsp-server
+
+          # python run-time dependencies:
+          python-pkgs.requests
+          python-pkgs.types-requests
+          python-pkgs.websockets
+        ]
+      );
+in
+pkgs.stdenv.mkDerivation {
```

Comment:
how does `mkDerivation` differs from `pkgs.mkShell`? shouldn't we use the latter in `shell.nix` files?

Reply:
i have the same question.

Reply:
`mkShell` is really only a thin wrapper around `stdenv.mkDerivation`, so in practice it often doesn't matter which you use:

https://github.com/NixOS/nixpkgs/blob/a3551324c60d8958e10c2479138fecbfa741e3d2/pkgs/build-support/mkshell/default.nix

The one nice thing that `mkShell` provides is the `inputsFrom` argument, which lets you easily(?) compose together these types of shells.  But if you're not using that argument, it normally doesn't really matter whether you use `mkShell` or `stdenv.mkDerivation`.

Until quite recently, `mkShell` would actually fail if you tried to build it directly:

https://github.com/NixOS/nixpkgs/commit/1e910209ae41475#diff-daf74c6dd39b73eb65b3102af73ee51666829e0a335c43acff28616b9bc964f4L48

This means that you weren't able to directly pass the output derivation from `mkShell` to `nix-build` in order to make sure it builds in CI.  Because of this, it was more common to see `stdenv.mkDerivation` used instead of `mkShell`.

This restriction was lifted a few years ago (in the above commit), so I imagine `mkShell` will become more widely used now.